### ISSUE TITLE
SvelteKit + Electron = 💔

### ISSUE THREAD

**Issue by jerrygreen** [sveltejs/kit#11997]
### Describe the bug

There's an issue when working from within ElectronJS, as there are not web-based routes, but instead file-based.

For starters, there's one issue inside generated `index.html` file, there's a line like:

```js
__sveltekit_wgqb1w = {
  base: new URL(".", location).pathname.slice(1, -1)
};
```

On Windows, it returns url something like that:

```js
'/C:/Users/Jerry/projects/electron-svelte/out/app.asar/.vite'
```

So the fact that it starts from `/C:/` is already an issue since it then return an error:

```txt
app.qFRTHcTk.js:7 Error: Not found: /C:/Users/Jerry/projects/electron-svelte/out/app.asar/.vite/index.html
```

I tried to intervene into Svelte building pipeline, so when building I make corresponding changes:

```diff
-__sveltekit_wgqb1w = {
-  base: new URL(".", location).pathname.slice(0, -1)
-};
+__sveltekit_wgqb1w = {
+  base: new URL(".", location).pathname.slice(1, -1)
+};
```

Now the url is formed properly, but now I'm stuck into some infinite loop by some reason:

```txt
Throttling navigation to prevent the browser from hanging. See https://crbug.com/1038223. Command line switch --disable-ipc-flooding-protection can be used to bypass the protection
```

It tries to load some initial code it seems, but I'm not sure where and why – since no error occurs. Only a warning:

![image](https://github.com/sveltejs/kit/assets/13215662/720c422e-f9b7-49ac-b5fa-6cd2294baf87)

There's a lot of blinking involved, so I can say myself: the thing is looped. It's broken. And it doesn't render any html as you can see.

Though if I'm not running from statics, but from web version like this:

```js
mainWindow.loadURL('http://localhost:5173')
```

Then it works perfectly fine:

![image](https://github.com/sveltejs/kit/assets/13215662/acf117ac-e4db-4ccd-ac07-02dcf7c7de53)

So the issue only seems related to when running from local static files like this:

```
mainWindow.loadFile(path.resolve(app.getAppPath(), '.vite/index.html'))
```

### Reproduction

Largely I use this minimal `electron-forge` repo that explains how to use electron with esm:

https://github.com/jdms754/electron-forge-vite-esm-test

I cloned it, added Svelte, and made it work.

It works perfectly for dev and production: i.e. both from `mainWindow.loadUrl` and `mainWindow.loadFile`.

And when I add Svelte into it, it works... But only in dev environment, i.e. `mainWindow.loadUrl`.

When it gets to building an executable with statics, I need to load statics like this:

```
mainWindow.loadFile(path.resolve(app.getAppPath(), '.vite/index.html'))
```

Then here, the problem occurs. All as described above: first with wrong base path, then with infinite loop.

### System Info

```Shell
Windows 10 (22H2)
Node: 20.11.1
Electron: 29.1.4
@sveltejs/adapter-auto: ^3.0.0 => 3.1.1
@sveltejs/adapter-static: 3.0.1 => 3.0.1
@sveltejs/kit: ^2.0.0 => 2.5.3
@sveltejs/vite-plugin-svelte: ^3.0.0 => 3.0.2
svelte: 5.0.0-next.80 => 5.0.0-next.80
vite: ^5.0.3 => 5.1.5
```

### Severity

blocking all usage of SvelteKit

**Comment by jerrygreen** [comment]
My guess, something is wrong here, starting at `if (hydrate) ...`:

![image](https://github.com/sveltejs/kit/assets/13215662/6dbbf141-bd99-4d22-8141-54c1527175e5)

It seems it constantly does `goto(location.href...` and that's the reason of infinite loop.

I'm not too much into the codebase, I don't know all details, but on a glance I can't even imagine how it even works in other use-cases, other than `file://` protocol. It almost seems like it should go into infinite loop in every single use-case because it always passes exactly two arguments to `kit.start(app, element)`, without hydrate argument, where at this point it's looped:

![image](https://github.com/sveltejs/kit/assets/13215662/9c173fc1-6f56-477e-877e-ef2c29a3def7)

There's obviously some magic happening that I don't see, because it's not infinitely looped in other cases than `file://` protocol.

---

I'm sorry for pinging but it seems quite important one, @Rich-Harris can you give some feedback? Like your thoughts of how/where it might go wrong that it goes into infinite loop? Or even more broadly: if you ever plan/want SvelteKit to support when used in conjunction with Electron? Or how the hydration works here. Any feedback appreciated, really. Thanks.

**Comment by khromov** [comment]
As far as I know SvelteKit apps have *never* worked when you run them from the disk, you always need a web server of some sort. Have you checked any of the available templates such as https://github.com/FractalHQ/sveltekit-electron ?

**Comment by jerrygreen** [comment]
For bundling, they use `electron-serve`, which is basically a file server.

Still, I'd prefer to run statics without any http wrapper.

I am not sure how viable this idea of running statics as is. Maybe it's not viable at all.
I'm sorry if I'm proposing some dumb idea that would take eternity to refactor SvelteKit just to this need.

And, actually, thank you for feedback. I will consider their approach as workaround, maybe it's not that bad as I think.

**Comment by khromov** [comment]
I think this limitation is mostly around the fact that JS modules (which SvelteKit uses heavily) can't be loaded from `file://` urls:

> You need to pay attention to local testing — if you try to load the HTML file locally (i.e. with a file:// URL), you'll run into CORS errors due to JavaScript module security requirements. You need to do your testing through a server.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_standard_scripts

There is a Vite plugin called `vite-plugin-singefile` that is able to wrap all the CSS and JS into a single file, this *can* run from a single index.html file. You can use this in Svelte, but I haven't seen anyone have luck with this in SvelteKit (probably due to the complicated nature of SK bundling)

**Comment by jerrygreen** [comment]
It’s no surprise `file://` protocol protected by cors, it wouldn’t be safe for browsers to allow requesting `file://` urls from other files: a lot of nasty stuff could be done just from opening a local html page, or worse – from a web page. Web security standards is on high stakes.

But electron is another story. It isn't just a web app, it's an executable app only then a web app, – and for executables it's already assumed they can do much more than simple html file, including file access. Thus, security standards are different. Chrome engine allows to pass `--allow-file-access-from-files` which negates cors issue for file access. And in Electron, it seems to be enabled by default. So, from an Electron app, I can fetch via `file://` protocol no problem.

Here's my little experiment in devtools, inside of electron app:

![image](https://github.com/sveltejs/kit/assets/13215662/e7aaf632-4976-4cbf-a9de-c121e984089d)

As you can see, I was able to access `index2.html` from `index1.html` both from `file://` protocol.

It works both for relative paths, and for absolute paths like:
```js
await fetch('file:///C:/Users/Jerry/projects/electron-template/.vite/renderer/index2.html')
await fetch('index2.html') // equivalent to this
```

So, this shouldn't be a problem to JS modules. Might require some code changes here and there in SvelteKit, still. Because currently SvelteKit fully assumes it is run from http. But it doesn't have to. I'd love to have Electron as first-class citized in SvelteKit.

**Comment by mhkeller** [comment]
@jerrygreen try setting `ssr = true` in your layout config: https://github.com/mhkeller/sveltekit-electron/

**Comment by w3rafu** [comment]
> @jerrygreen try setting `ssr = true` in your layout config: https://github.com/mhkeller/sveltekit-electron/

This doesn't work.

**Comment by HuakunShen** [comment]
I had problem loading sveltekit in Electron on Mac (svelte with vite works).
![image](https://github.com/user-attachments/assets/55c530cc-46d1-4030-bdf7-a9bd0e4a9f59)
The error is also related to goto 
![image](https://github.com/user-attachments/assets/fadd8e58-6bab-4641-9cf8-a0b69428ffec)

**Comment by eltigerchino** [comment]
I've create a minimal working template after digging around a bit. It can probably be improved by loading from the Vite dev server URL during development.
https://github.com/eltigerchino/sveltekit-electron-template

In the template, we're building an SPA and loading the build output `index.html` file using [`electron-serve`](https://github.com/sindresorhus/electron-serve/tree/main).

We can't use electron's `mainWindow.loadFile` because that serves it using the `file://` protocol which has its limitations. One of them being that it causes Electron to serve `index.html` with a URL that is incompatible with the SvelteKit pathname router. e.g., it sets the window URL to an absolute path like `file:///foo/bar/build/index.html` but navigating to `/about` would change the URL to `file:///about` which Electron intercepts and complains that it can't find the file at that path. `electron-serve` solves this problem by using a custom protocol such as `app://` and setting the window URL to `app://-/` which would correctly navigate to `app://-/about` (`-` is the default origin).

You could probably do without `electron-serve` and load the `index.html` file with `mainWindow.loadFile` if you used SvelteKit's new [hash router](https://svelte.dev/docs/kit/configuration#router) since it would not cause the URL pathname to change.

**Comment by jerrygreen** [comment]
> We can't use electron's mainWindow.loadFile because that serves it using the file:// protocol which has its limitations. One of them being that it causes Electron to serve index.html with a URL that is incompatible with the SvelteKit pathname router

Yeah, that’s the main claim here.

Using electron-serve, it was possible to do long ago, but it includes server into the app, which already has entire browser, – all of this makes these web technologies even more bloatware than they already are.

I guess it is the only way it can currently work 🤷‍♂️ other ways would require enormous work, it would require doing custom adapter for electron or something.

**Comment by jsingerlenovo** [comment]
I too am trying to get a static build of SvelteKit to work with Electron's `loadFile()` (technically `loadURL()` with `protocol: "file"`).

First, I ran into the problem of **modulepreload** (`<link rel="modulepreload" href=...>`) causing CORS issues.  

1. I wondered if I could turn that off by using `<script src=...>` instead since scripts are grandfathered in with no CORS issues by changing `config.kit.output.preloadStrategy` but `'none'` or `'script'` doesn't seem to be an option. ❌
2. I found a workaround by using `config.kit.output.bundleStrategy: 'inline'` so that no modulepreloads would be generated. ✅ Though it would be nice if 'inline' wasn't required.

Second, I ran into the problem of **routing with the pathname** because the pathname would be a full path like `'/C:/Program%20Files%20(x86)/...<cut-for-brevity>.../app.asar/build/ui/index.html'`.

1. This causes `get_navigation_intent()` to look for a path of `'/index.html'` rather than simply `'/'`.
2. That causes `get_navigation_intent()` to not find the path in its list of `routes`, which causes `intent` to be `undefined`.
3. That causes `navigation_result` to be `undefined` which causes us to fall into ``await handle_error(new SvelteKitError(404, "Not Found", `Not found: ${url.pathname}`), {...})``.
4. Therefore, it's not that it can't actually find the HTML file, it's that it can't find the `'/index.html'` route, because the real route is `'/'` in the list of `routes`. ❌

Other things I noticed:

- I saw that after this it falls down to `const updated2 = await stores.updated.check();` causing a fetch on **version.json** (``const res = await fetch(`${assets}/${"_app/version.json"}`, {...})``) which fails due to CORS.
    -  I tried setting `config.kit.version.pollInterval: 0` but that doesn't stop this `stores.updated.check()` call, it only stops it from regularly polling.
- I've noticed that when using `config.kit.router.type: 'pathname'` the `base` in some configurations would be empty which affects `get_url_path()` (though I'd need to double-check that to be sure).

Unfortunate workaround:

To solve this, I've turned on `config.kit.router.type: 'hash'` and **hash** handles routing correctly ✅ but has the unfortunate side-effect of changing all of our anchor tags and goto() calls to using `'#/routeName'` rather than just `'/routeName'`. ❌

In summary: 

We need a way for the path of `'/index.html'` to match the `'/'` route.  (And to not fetch version.json.)

**Comment by khromov** [comment]
@jsingerlenovo What I found when experimenting with this is that version.json is only requested if the route does not match anything, try creating a `/src/routes/[[catch]]` route and see if that helps you.

**Comment by jsingerlenovo** [comment]
@khromov Thanks for the quick reply!  And great suggestion!  I tried creating:

1. `/src/routes/[[catch]]` folder, nothing changed (no routes added, version.json still fetched).
2. `/src/routes/[[catch]]/+page.svelte` file, I got a build error `The "/" and "/[[catch]]" routes conflict with each other`.
3. `/src/routes/[[...catch]]/+page.svelte` file, I got a build error `Invalid route /[[...catch]] — a rest route segment is always optional, remove the outer square brackets`.
4. `/src/routes/[...catch]/+page.svelte` file, this created a new `'/[...catch]'` route in the `routes`, which avoided fetching version.json like you said!  

That's progress! ✅  

Now, if on that catch-all page `src/routes/[...catch]/+page.svelte`, I have links and a goto() to other pages, for example:
```
<script lang="ts">
    import { goto } from "$app/navigation";

      function navigateTo() {
          goto("/chat");
      }
  </script>

<h1>Catch All Page</h1>

<a href="/">Home</a>
<br />

<button onclick={() => navigateTo()}>Go to Chat Page</button>
<br />

<a href="/chat">Link to Chat Page</a>
```

then clicking on any of those goes to the wrong URL: ❌
1. Clicking on `Home` - `'/'` causes the browser to go to file:///C:/
2. Clicking on `Go to Chat Page` - `goto("/chat")` causes the browser to go to file:///C:/chat
3. Clicking on `Link to Chat Page` - `'/chat'` causes the browser to go to the same file:///C:/chat

However, if I run `npm run preview`, every click goes to the correct page.

**In summary**: We now seem to have a navigation problem.

Looking in `resolve_url()`, I see:
1. `url` coming in as `"/chat"`
2. `baseURI` gets set to the full path like `"file:///C:/myapp/build/index.html"`
3. then we do `return new URL(url, baseURI);`

This works great with regular URLs:
```
new URL("/chat", "https://svelte.dev")
// href: "https://svelte.dev/chat"
// origin: "https://svelte.dev"
// pathname: "/chat"
```

but not for file://
```
new URL("/chat", "file:///C:/myapp/build/index.html")
// href: "file:///C:/chat"
// origin: "file://"
// pathname: "/C:/chat"
```

It seems to truncate the `origin` down to just the `"file://"` and adds `"/C:"` to the _beginning_ of `pathname`.

Maybe we need something like:
```
if (baseURI.protocol === 'file:') {
    return new URL(baseURI.href.substring(0, baseURI.href.lastIndexOf('/')) + path + (baseURI.href.endsWith('.html') ? '.html' : ''));
}
else {
    return new URL(url, baseURI);
}
```

or just use a regex:
```
if (baseURI.protocol === 'file:') {
    return new URL(baseURI.href.replace(/\/?[^/]*?(\.html)?$/, path + "$1"));
}
else {
    return new URL(url, baseURI);
}
```

which turns `"file:///C:/myapp/build/index.html"` into `"file:///C:/myapp/build/chat.html"`
or turns `"file:///C:/myapp/build/index"` into `"file:///C:/myapp/build/chat"`

Think this would work?  I'm not sure if there are other places in the code or other scenarios that need to be considered.

**Comment by khromov** [comment]
My mistake @jsingerlenovo , I meant `[...catch]` not `[[catch]]`:
https://svelte.dev/docs/kit/advanced-routing#Rest-parameters

It turns out if you set the router type to `type: "hash",` you can also use normal routes (not a catch route). Here is a sample project that uses normal routing and works by just double clicking the resulting `index.html` file:
https://github.com/khromov/sveltekit-file-routing

I **do not think** you will be able to get "normal" routing working, you'll have to use hash routing, eg `goto("#/chat")` instead of `goto("/chat")`. I think this is inherently not a SvelteKit limitation but a limitation in how the `file://` protocol works.

**Comment by jsingerlenovo** [comment]
Thanks again, @khromov !  I cloned your repo and took a look.  Looks like hash-based routing is best choice right now for Electron.

I think with the following changes, **SvelteKit could enable pathname-based routing even with the `file://` protocol for Electron**, as long as `goto()` is always used (normal anchor tags do not seem to work).

1. Change `get_url_path()` from:
```
function get_url_path(url) {
	return (
		decode_pathname(
			app.hash ? url.hash.replace(/^#/, '').replace(/[?#].+/, '') : url.pathname.slice(base.length)
		) || '/'
	);
}
```

to:

```
function get_url_path(url) {
	let newPathname = '';
	if (app.hash) {
		newPathname = url.hash.replace(/^#/, '').replace(/[?#].+/, '')
	}
	else {
		newPathname = url.pathname.slice(base.length)
	}

	const decodedPathname = decode_pathname(newPathname) || '/';

	if (url.protocol === 'file:') {
		if (decodedPathname.endsWith('/index.html')) {
			return decodedPathname.slice(0, -('index.html'.length));
		}
		else if (decodedPathname.endsWith('.html')) {
			return decodedPathname.slice(0, -('.html'.length));
		}
	}
	return decodedPathname;
}
```

2. Change `navigate()` from:
```
// ...
		fn.call(history, entry, '', url);
// ...
```

to:
```
// ...
		// I don't know if there is a way to fix this, but for now I simply removed this call 
		// for file:// urls because I was getting "Uncaught (in promise) SecurityError: Failed 
		// to execute 'pushState' on 'History': A history state object with URL 
		// 'file:///C:/git/myapp/build/chat.html' cannot be created in a document with 
		// origin 'null' and URL 'file:///C:/git/myapp/build/index.html'.

		if (url.protocol !== "file:") {
			fn.call(history, entry, '', url);
		}
// ...
```

3. Change `resolve_url()` from:
```
export function resolve_url(url) {
	if (url instanceof URL) return url;

	let baseURI = document.baseURI;

	if (!baseURI) {
		const baseTags = document.getElementsByTagName('base');
		baseURI = baseTags.length ? baseTags[0].href : document.URL;
	}

	return new URL(url, baseURI);
}
```

to:
```
export function resolve_url(url) {
	if (url instanceof URL) return url;

	let baseURI = document.baseURI;

	if (!baseURI) {
		const baseTags = document.getElementsByTagName('base');
		baseURI = baseTags.length ? baseTags[0].href : document.URL;
	}

	const baseUrl = new URL(baseURI);
	if (baseUrl.protocol === 'file:' && url.startsWith("/")) {
		return new URL(baseUrl.href.replace(/\/?[^/]*?(\.html)?$/, url + "$1"));
	}
	else {
		return new URL(url, baseURI);
	}
}
```

To be clear, I did not write this for production, just trying to see if it could work, thus it is not a PR, it could likely be improved and needs tested more thoroughly.  However...

Simply opening index.html in a browser with file:// **I can click on a button that uses goto('/chat') or goto('/users') or goto('/') and it will take me to the correct pages!**  I did notice that the address bar in the browser never changes, but not an issue in Electron.

I leave this code here to hopefully jog someone's imagination to implement this more fully. 😊

**Comment by yeori** [comment]
It is achieved by implementing custom protocol in Electron.

[protocol](https://www.electronjs.org/docs/latest/api/protocol)

The key is to replace `file://app_document_root/uri/to/resource` with  `app://dummyhost/uri/to/resource` and then read resource from `app://dummyhost/uri/to/resource` to local file matching `/uri/to/resource`.

* `app` - scheme like http, https. you can name it freely.
* `dummyhost` - hostname like `localhost`, You can name it freely.

Protocol implementation can read file from your local on main process of Electron.

When building sveltekit app with command `npm run build`, it produces files 
```
./.svelte-kit/output
├── client
│   ├── _app
│   │   ├── immutable(js files here)
│   │   └── version.json
│   └── favicon.png
├── prerendered
│   ├── dependencies
│   │   └── _app
│   └── pages
│       └── index.html
```
`dist/index.html` is like this.
```html
<html lang="en">
	<head>
		<meta charset="utf-8" />
		<link rel="icon" href="./favicon.png" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		
		<link rel="modulepreload" href="./_app/immutable/entry/start.BJ2uKHWk.js">
		<link rel="modulepreload" href="./_app/immutable/chunks/Da7Zbzlw.js">
		...
	</head>
	<body data-sveltekit-preload-data="hover">	
			<script>
				{
					Promise.all([
						import("./_app/immutable/entry/start.BJ2uKHWk.js"),
						import("./_app/immutable/entry/app.DcuRZulF.js")
					])
				}
			</script>
	</body>
</html>
```
I copied `client/_app` and `prerender/page/index.html` into the `renderer` directory of Electron like this.
```
./src/renderer
├── _app
│   ├── immutable
│   │   ├── chunks
│   │   │   ├── 26I3BnCA.js
│   │   │   ├── BdGLW2Tw.js
│   │   │   ├── DSYY6uCT.js
│   │   │   ├── Da7Zbzlw.js
│   │   │   ├── Zg7hjcAr.js
│   │   │   └── nwBLtpdx.js
│   │   ├── entry
│   │   │   ├── app.DcuRZulF.js
│   │   │   └── start.BJ2uKHWk.js
│   │   └── nodes
│   │       ├── 0.DN3w3FhW.js
│   │       ├── 1.Bs7BEbGQ.js
│   │       └── 2.3UYzGWBf.js
│   └── version.json
└── index.html
```

### electron.vite.config.js

All files under the `./src/renderer` are already built from sveltekit, so they do not have to be built again.

```ts
import extra from 'fs-extra'

const { copySync } = extra
export default defineConfig({
  main: {
    ...
  },
  preload: {
    ...
  },
  renderer: {
    build: {
      write: false
    },
    plugins: [
      {
        name: 'copy-renderer',
        buildEnd: (): void => {
          copySync('src/renderer', 'out/renderer', { overwrite: true })
        }
      }
    ]
  }
})
```
* build.write: false(prevents index.html to be built)
* copySync - `npm install fs-extra`.

Running `npm run build` from Electron app produces output.
```
./out
├── main
│   └── index.js
├── preload
│   └── index.mjs
└── renderer
    ├── _app
    │   ├── immutable
    │   │   ├── chunks
    │   │   │   ├── 26I3BnCA.js
    │   │   │   ├── BdGLW2Tw.js
    │   │   │   ├── DSYY6uCT.js
    │   │   │   ├── Da7Zbzlw.js
    │   │   │   ├── Zg7hjcAr.js
    │   │   │   └── nwBLtpdx.js
    │   │   ├── entry
    │   │   │   ├── app.DcuRZulF.js
    │   │   │   └── start.BJ2uKHWk.js
    │   │   └── nodes
    │   │       ├── 0.DN3w3FhW.js
    │   │       ├── 1.Bs7BEbGQ.js
    │   │       └── 2.3UYzGWBf.js
    │   └── version.json
    └── index.html
```

### ProtocolHandler

A protocol handler should parse request.url into local file.
If req.url is `app://internal/index.html`, it should read file `./out/renderer/index.html`.

**Comment by jerrygreen** [comment]
> It is achieved by implementing custom protocol in Electron.
> @yeori 

Honestly it seems to be quite good of a solution. It still has some overhead because it is basically a glue transforming custom protocol urls to local file structure, it doesn’t make useful calculations other than just to make the thing working, doesn’t solve the bug but works around it. Ideally, the app itself should just fetch files from file protocol as it already knows what it needs and such glue is fundamentally not necessary. But unlike the solutions above posted by others, at least this solution doesn’t require additional node process, so it’s minuscule overhead compared to others. It’s looking quite good.

I’m not closing the issue because I think there is still a better solution where SvelteKit gets fixed and it doesn’t infinitely loop in cases where Electron starts from file protocol. One might even think it’s a bug with Electron, but I believe it’s SvelteKit isn’t best adapted for such use-case.

In meanwhile, your workaround looks best. If I ever return to an idea of making SvelteKit+Electron app, I will definitely try this. Thanks for posting this ❤️

**Comment by CPritch** [comment]
I'm using this package and it works great for me when packaged https://www.npmjs.com/package/electron-serve. Basically just dresses up the protocol interception.

**Comment by teemingc** [comment]
@LukeHagar made an adapter for electron applications https://github.com/LukeHagar/sveltekit-adapters#electron-adapter hope folks here can give it a try

**Comment by LukasParke** [comment]
Howdy!

Yes the adapter linked above uses the same principal as outlined in some of the other conversations above. 

The key difference though is that I adapted the core handle logic from the node adapter and hooked it up to an electron protocol handler for `http` traffic.

This means that you are able to build a SvelteKit app as normal, simply include a `main.(js|ts)` and `preload.(js|ts)` for any logic specific to those files, things like the size, styling, and behavior of the main window, or any specific things you want to use IPC for, and the adapter bundles everything up into a final output folder that is ready to be built with `electron-builder`.

Now a VERY exciting thing to note here, is that this adapter includes full support for the backend functionality you would hope to see in a fully functional SvelteKit application, your SvelteKit "Server" code is capable of interacting with the local filesystem or any other special Electron APIs however you would like, and your Client code functions just like normal. 

All of the assets for the app are served with the `_app` URLs the same way they would be with any other deployment method. 

I tested everything pretty thoroughly while developing it, and I used the Sverdle example as a reference to make sure things like form actions worked as expected and had everything to a point I was really happy with.

That being said I certainly could have missed something :D, so if anyone has any notes, or issues, feel free to open an issue [here](https://github.com/LukeHagar/sveltekit-adapters/tree/main/packages/adapter-electron) and I'll do my best to assist.

Thanks for sharing that @eltigerchino :)

**Comment by webJose** [comment]
I understand the OP:  He wants to use Sveltekit because he likes its router.  No problem.

If anyone of you out there would be willing to drop Sveltekit and use just Svelte + my router, my router works in Electron:  [@wjfe/n-savant](https://github.com/WJSoftware/wjfe-n-savant).

The online doc that explains the needed trick:  [Electron Support](https://wjfe-n-savant.hashnode.space/wjfe-n-savant/introduction/electron-support)

> [!IMPORTANT]
> I only tested in a Windows PC.  I don't have a Linux or Mac PC to test on.  **I depend on your feedback** to know:  [Please provide feedback here](https://github.com/WJSoftware/wjfe-n-savant/discussions/57).  Thank you!

**Comment by LukasParke** [comment]
> I understand the OP:  He wants to use Sveltekit because he likes its router.  No problem.
> 
> If anyone of you out there would be willing to drop Sveltekit and use just Svelte + my router, my router works in Electron:  [@wjfe/n-savant](https://github.com/WJSoftware/n-savant).
> 
> The online doc that explains the needed trick:  [Electron Support](https://wjfe-n-savant.hashnode.space/wjfe-n-savant/introduction/electron-support)
> 
> > [!IMPORTANT]
> > I only tested in a Windows PC.  I don't have a Linux or Mac PC to test on.  **I depend on your feedback** to know:  [Please provide feedback here](https://github.com/WJSoftware/wjfe-n-savant/discussions/57).  Thank you!

I would think people would be more inclined to use the first party solution.

It's not JUST routing that SvelteKit gives you, and the solution above worked VERY well for me.

**Comment by ManaHime** [comment]
> I understand the OP: He wants to use Sveltekit because he likes its router. No problem.
I like Sveltekit's router sure, but I also like all the other things the Sveltekit gives me.

> If anyone of you out there would be willing to drop Sveltekit and use just Svelte + my router, my router works in Electron: [@wjfe/n-savant](https://github.com/WJSoftware/n-savant).
Honestly at the end of the day if I can have a solution that allows me to build an entire app without having to juggle various different system and keep the same dev feeling across the entire stack, then I'd rather do that.

Incidentally the electron adapter linked above is probably as close as I can probably ever hope to get. It does look very promising.

Also If the Svelte team ever makes something that makes building apps easier with Svelte than I'm all for it. ＾＾

**Comment by webJose** [comment]
@ManaHime I just realized the link to the GH repo was broken.  Here it is, just in case:  https://github.com/WJSoftware/wjfe-n-savant

**Comment by MichaelJCole** [comment]
Guys, I was so excited to get into sveltekit and try it out for my electron app.  But I'm stuck on setting it up to get started.  

Nuxt has @nuxt/electron which is baller: https://nuxt.com/modules/electron

It uses vite-plugin-electron:  https://github.com/electron-vite/vite-plugin-electron

Isn't SSR worse performance on the same CPU than an SPA build (more IO)?   
If using electron's IPC, don't I just want HMR and hot reload?

Is there anything I'm missing? Thnx!

**Comment by teemingc** [comment]
> Guys, I was so excited to get into sveltekit and try it out for my electron app. But I'm stuck on setting it up to get started.
> 
> Nuxt has @nuxt/electron which is baller: [nuxt.com/modules/electron](https://nuxt.com/modules/electron)
> 
> It uses vite-plugin-electron: [electron-vite/vite-plugin-electron](https://github.com/electron-vite/vite-plugin-electron)
> 
> Isn't SSR worse performance on the same CPU than an SPA build (more IO)? If using electron's IPC, don't I just want HMR and hot reload?
> 
> Is there anything I'm missing? Thnx!

You can give https://github.com/LukeHagar/sveltekit-adapters#electron-adapter a try
